lib/OpenQA/Schema/Result/JobModules.pm

```diff
@@ -137,13 +139,24 @@ sub results {
     for my $step (@$details) {
         my $text_file_name = $step->{text};
         if (!$skip_text_data && $text_file_name && !defined $step->{text_data}) {
-            my $file = path($dir, $text_file_name);
+            my $file      = path($dir, $text_file_name);
+            my $try_again = 0;
             try {
-                $step->{text_data} = decode('UTF-8', $file->slurp) if -e $file;
+                $step->{text_data} = decode('UTF-8', $file->slurp);
             }
             catch {
-                $step->{text_data} = "Unable to read $text_file_name.";
+                my $error = $_;
+                # try reading the results one more time if the JSON file's size has increased
+                # note: Likely a concurrent finalize_job_results Minion job has finished so the separate text file has
+                #       just been incorporated within the JSON file.
+                if (!$text_data_read_attempt && ($error =~ qr/No such file/) && ((-s $fn // -1) > $initial_file_size)) {
+                    $try_again = 1;
+                }
+                else {
+                    $step->{text_data} = "Unable to read $text_file_name.";
+                }
             };
+            return $self->results(%options, text_data_read_attempt => 1) if $try_again;
         }
 
```

Comment:
Just adding `return $self->results(%options) if -s $fn > $initial_file_size;` after the loop would be much cleaner. You don't even need `text_data_read_attempt` because the JSON file will change at most once.

Reply:
I simplified it a little bit.